segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/ContainerEventProcessorImpl.java

```diff
@@ -0,0 +1,523 @@
+                        }
+                        return null;
+                    });
+        }
+
+        /**
+         * Attempts to read data from the internal Segment. The size of the read length would be the minimum between the
+         * outstanding bytes and 2 * ProcessorEventData.MAX_TOTAL_EVENT_SIZE bytes.
+         *
+         * @return A {@link CompletableFuture} that, when completed, will contain a {@link EventsReadAndTruncationLength}
+         * object with events read and the next truncation length for the internal Segment.
+         */
+        private CompletableFuture<EventsReadAndTruncationLength> readEvents() {
+            return CompletableFuture.supplyAsync(() -> {
+                                        int readLength = (int) Math.min(getOutstandingBytes(), 2 * ProcessorEventData.MAX_EVENT_SIZE);
+                                        return this.segment.read(this.segment.getInfo().getStartOffset(), readLength, containerOperationTimeout);
+                                    }, this.executor)
+                                    .thenCompose(rr -> AsyncReadResultProcessor.processAll(rr, this.executor, containerOperationTimeout))
+                                    .thenApply(this::deserializeEvents);
+        }
+
+        @SneakyThrows(Exception.class)
+        private EventsReadAndTruncationLength deserializeEvents(BufferView inputData) {
+            List<ProcessorEventData> events = new ArrayList<>();
+            long truncationLength = 0;
+            Exception deserializationException = new NoDataAvailableException();
+            try {
+                @Cleanup
+                BoundedInputStream input = new BoundedInputStream(inputData.getReader(), inputData.getLength());
+                int dataLength = inputData.getLength();
+                while (input.getRemaining() > 0 && events.size() < this.config.getMaxItemsAtOnce()) {
+                    ProcessorEventData event = this.serializer.deserialize(input);
+                    events.add(event);
+                    // Update the truncation length, which includes user data and internal VersionedSerializer metadata.
+                    truncationLength = dataLength - input.getRemaining();
+                }
+            } catch (BufferView.Reader.OutOfBoundsException ex) {
+                // Events are of arbitrary size, so it is quite possible we stopped reading in the middle of an event.
+                // Only if we have already read some data, we do not throw.
+            } catch (Exception ex) {
+                // Some unexpected serialization error occurred here; rethrow if this prevents reading further events.
+                log.error("{}: Problem while deserializing events.", this.traceObjectId);
+                deserializationException = ex;
+            }
+
+            // If no events have been read, throw the right exception. Otherwise, process whatever events have been read.
+            if (events.isEmpty()) {
```

Comment:
So what if you have a deserialization exception but also read some events? Is anyone going to know about that? I guess yes, because you log it, but then that event will likely be lost. How is your code handling this situation?

Reply:
Yes, we have discussed that already, and I can summarize the intended behavior as follows:
- If no exception is caught and some events are read, just proceed processing the events.
- If an exception is caught and we have successfully read some events:
  - Expected exception (e.g., partial event read): just process the events and truncate.
  - Unexpected exception (e.g., serialization issue): log the error, process the events and truncate.
- If an exception is caught and no events are read:
  - Expected exception (e.g., partial event read): throw `NoDataAvailableException`, which is expected in the processing iteration.
  - Unexpected exception (e.g., serialization issue): log the error details and throw that exception upstream, interrupting the processing iteration at that point.

So, if the error is transient, we will log it but process until the last valid event. If the error is persistent, any successive iteration will get interrupted as the exception will be re-thrown.